examples/http-server/src/Main.gren

```diff
@@ -0,0 +1,129 @@
+        }
+
+
+type alias Model =
+    { stdout : Stream
+    , stderr : Stream
+    , server : Maybe Http.Server
+    }
+
+
+type Msg
+    = CreateServerResult (Result Http.CreateServerError Http.Server)
+    | GotRequest Http.Request Http.Response
+    | ResponseFinished Http.Response
+
+
+init : Program.AppInitTask { model : Model, command : Cmd Msg }
+init =
+    Program.await Node.initialize <| \nodeConfig ->
+        Program.await Http.initialize <| \serverPermission ->
+            Program.startProgram
+                { model =
+                    { stdout = nodeConfig.stdout
+                    , stderr = nodeConfig.stderr
+                    , server = Nothing
+                    }
+                , command =
+                    Task.attempt CreateServerResult <|
+                        Http.createServer serverPermission
+                            { host = "localhost"
+                            , port_ = 3000
+                            }
+                }
+
+
+update : Msg -> Model -> { model : Model, command : Cmd Msg }
+update msg model =
+    case msg of
+        CreateServerResult result ->
+            case result of
+                Ok server ->
+                    { model = { model | server = Just server }
+                    , command = Stream.sendLine model.stdout
+                        "Server started"
+                    }
+                Err _ ->
+                    { model = model
+                    , command = Stream.sendLine model.stderr
+                        "Server failed to start"
+                    }
+        
+        GotRequest req res ->
+            let
+                body = case { method = req.method, path = req.url.path } of
+                    { method = GET, path = "/hello" } ->
+                        "Hello to you too!"
+                    
+                    { method = POST } ->
+                        "You posted: " ++ req.body
+                    
+                    _ ->
+                        Http.requestInfo req
+            in
+            { model = model
+            , command =
+                Http.respondWithString res
+                    { status = 200
+                    , body = "<html>" ++ body ++ "</html>"
+                    , headers = Dict.fromArray
+                        [ { key = "Content-type", value = "text/html"}
+                        ]
+                    }
+                    |> Task.perform ResponseFinished
+            }
+        
+        -- TODO: idea for future API....
+        --     { model = model
+        --     , command = Cmd.batch
+        --         [ Stream.sendLine model.stdout <| Debug.toString req.url
+        --         , Http.response req
+        --             |> Response.status 200
+        --             |> Response.header "Content-type" "text/html"
+        --             |> Response.chunk "<html>Hello from gren!" (SentChunk 1)
+        --             |> Response.chunk "</html>" (SentChunk 2)
```

Comment:
I think this future API is elegant.

The one thing I would change is that `SentChunk`/`ResponseFinished` is turned into subscriptions. That way, they are optional.

Reply:
Could use some advice on the mental model around interactions with the Kernel and scheduler. Is it ok for `update` to call functions that mutate an object in the kernel js (e.g. `Response.chunk` results in an immediate call to `response.write`)? or is that considered a side effect that should be handled with a task (hence my messages here)?

Reply:
That would be a side-effect, yeah.

I think the `Response.x` functions should build up a record with what you intend to do, then have a `Response.send` (or something) that takes the "intent record` and returns a task which when run will actually perform all the requested things.

Reply:
That makes sense, thanks.

I still have a question about `ResponseFinished` being optional: If `Response.send` (or something) returns a task, won't it require having a message like that? unless it returns a command instead?

meaning it would have look something like this (still playing with names):

```elm
...
|> Response.write "something"
|> Response.write "something"
|> Response.send ResponseSent
```

?

Reply:
Interesting question!

So, one trick that is used in several places in Elm, is to return a generic `Cmd` as a way to avoid sending in a `Message` to the API. This naturally only works if a message won't be sent to the `update` function.

So, depending on what is useful, you could have

```
send : ResponseIntent -> Task ResponseFailed {}
```

This will allow you await the sent response and perform stuff like logging, registering metrics etc.

_or_

```
send : ResponseIntent -> Cmd a
```

Gren will figure out that `a` has to be the `Message` type, and so that should work. The downside of this is that you cannot chain things directly of sending a response (no way of awaiting commands). That, however, might not be a problem in an application that directly receives and sends responses. By that point you've already setup the TEA structure anyway, and if you need to do something once responses have been sent, you can use subscriptions for that purpose.

Reply:
I'm starting to realize that the focus for 0.5 (0.4 being parametric modules) should be tasks, commands, and processes 😅

Reply:
I've started making this the working API. Made the most sense as I started trying to work on the remaining TODOs. Hopefully have an update to the PR by tomorrow, but can't promise...